File: PHP/PenPaper/Paper.php
<?php
    require_once 'Exceptions/InvalidParam.php';
    require_once 'Exceptions/InvalidAttribute.php';
    require_once 'Exceptions/OutOfSpace.php';
    
    class Paper {
        private $symbols, $maxSymbols;
        private $content;
        
        private function validate($value) {
            if ( is_int($value) && $value > 0 ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        public function __construct($maxSymbols = 4096) {
            $this->maxSymbols = $this->validate($maxSymbols);
            $this->symbols = 0;
            $this->content = "";
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new InvalidAttribute("Attribute error: attribute '$name' not found.");
        }
        
        public function addContent($msg) {
            $this->symbols = strlen($this->content) + strlen($msg);
            
            if ( strlen($this->content) == $this->maxSymbols ) {
                throw new OutOfSpace("Not enough space! You've used all free space on the paper!");
            }
            if ( $this->symbols > $this->maxSymbols ) {
                $this->content .= substr($msg, 0, $this->maxSymbols - strlen($this->content));
                return;
            }
            $this->content .= $msg;
        }
        
        public function show() {
            echo $this->content . PHP_EOL;
        }
    }
?>

File: W5/Pen&Paper/Paper.py
from Exceptions import OutOfSpaceError, PaperMaxSymbolsError


class Paper:
    def __init__(self, max_symbols: int = 4096) -> None:
        if max_symbols <= 0:
            raise PaperMaxSymbolsError('You entered incorrect max symbols for paper!')
        self._max_symbols = max_symbols

        self._symbols = 0
        self._content = ''

    @property
    def symbols(self) -> int:
        return self._symbols

    @property
    def max_symbols(self) -> int:
        return self._max_symbols

    def add_content(self, msg: str) -> None:
        start = 0

        if self._max_symbols <= 0:
            raise OutOfSpaceError('Not enough space! You used all free space on the paper!')

        new_msg = msg[start:self._max_symbols]
        self._max_symbols -= len(new_msg)
        self._content += new_msg
        self._symbols = len(self._content)

    def show(self) -> None:
        print(self._content)
